exporter/googlecloudstorageexporter/exporter.go

```diff
@@ -5,43 +5,185 @@ package googlecloudstorageexporter // import "github.com/open-telemetry/opentele
 
 import (
 	"context"
+	"errors"
+	"fmt"
+	"net/http"
 
+	"cloud.google.com/go/compute/metadata"
+	"cloud.google.com/go/storage"
+	"github.com/google/uuid"
 	"go.opentelemetry.io/collector/component"
 	"go.opentelemetry.io/collector/consumer"
 	"go.opentelemetry.io/collector/exporter"
 	"go.opentelemetry.io/collector/pdata/plog"
 	"go.uber.org/zap"
+	"google.golang.org/api/googleapi"
 )
 
 type storageExporter struct {
-	// TODO
+	cfg           *Config
+	logsMarshaler plog.Marshaler
+	storageClient *storage.Client
+	bucketHandle  *storage.BucketHandle
+	logger        *zap.Logger
 }
 
 var _ exporter.Logs = (*storageExporter)(nil)
 
+func newGCSExporter(
+	ctx context.Context,
+	cfg *Config,
+	logger *zap.Logger,
+) (*storageExporter, error) {
+	return newStorageExporter(
+		ctx,
+		cfg,
+		metadata.ZoneWithContext,
+		metadata.ProjectIDWithContext,
+		logger,
+	)
+}
+
 func newStorageExporter(
-	_ context.Context,
-	_ *Config,
-	_ *zap.Logger,
-) *storageExporter {
-	// TODO
-	return &storageExporter{}
+	ctx context.Context,
+	cfg *Config,
+	getZone func(context.Context) (string, error),
+	getProjectID func(context.Context) (string, error),
+	logger *zap.Logger,
+) (*storageExporter, error) {
+	if cfg.Bucket.Region == "" {
+		var err error
+		if cfg.Bucket.Region, err = getZone(ctx); err != nil {
+			return nil, fmt.Errorf("failed to get region from metadata: %w", err)
```

Comment:
We should clarify these two error messages to the overall problem, which is that there is no region accessible. The fact that it failed from metadata is only part of the problem.

Reply:
I get it now. I changed the error message. Does it look good?